components/authz-service/server/v2/rules_property_test.go

```diff
@@ -18,17 +21,511 @@ import (
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[0])
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				updateReq := api.UpdateRuleReq{
+					Id:         reqs.rules[0].Id,
+					ProjectId:  reqs.rules[0].ProjectId,
+					Name:       reqs.rules[0].Name + " updated",
+					Type:       reqs.rules[0].Type,
+					Conditions: reqs.rules[0].Conditions,
+				}
+				rStaged, err := cl.UpdateRule(ctx, &updateReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				rApplied, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rApplied.Rule.Status == "applied" &&
+					ruleMatches(updateReq, *rApplied.Rule) &&
+					rStaged.Rule.Status == "" && // TODO: this is wrong; should be "staged"
+					ruleMatches(updateReq, *rStaged.Rule)
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("updates an applied rule and applies update confirming transition from applied to staged to applied",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[0])
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				rInitialApplied, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				updateReq := api.UpdateRuleReq{
+					Id:         reqs.rules[0].Id,
+					ProjectId:  reqs.rules[0].ProjectId,
+					Name:       reqs.rules[0].Name + " updated",
+					Type:       reqs.rules[0].Type,
+					Conditions: reqs.rules[0].Conditions,
+				}
+				rStaged, err := cl.UpdateRule(ctx, &updateReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				rFinalApplied, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rInitialApplied.Rule.Status == "applied" &&
+					rStaged.Rule.Status == "" && // TODO: this is wrong; should be "staged"
+					rFinalApplied.Rule.Status == "applied" &&
+					ruleMatches(updateReq, *rStaged.Rule) &&
+					ruleMatches(updateReq, *rFinalApplied.Rule)
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("updates an applied and staged rule and applies update confirming transition from staged to applied",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
```

Comment:
I think we've got these two function calls so often they might also be refactored into a little helper? I mean the "create project; create rule" steps.